lib/spack/spack/fetch_strategy.py

```diff
@@ -551,7 +551,12 @@ def fetch(self):
             os.remove(filename)
 
         # Symlink to local cached archive.
-        symlink(path, filename)
+        try:
+            symlink(path, filename)
+        except SymlinkError as e:
+            tty.warn("No-op as symlink to local cached archive was not created.")
+            tty.debug(e)
+            return
```

Comment:
The test failure that prompted this was `lib/spack/spack/test/cache_fetch.py::test_fetch`, I added a comment about what was happening to cause the failure:

```
    Since the `stage.save_filename` value is the same as the `cache` path, fetch()
    deletes the cache.tar.gz file, then tries to create a symlink that points to
    itself for the file that it just deleted. This raises a symlink.Symlink error
    because the resulting link is broken.
```

The `return` worries me now looking back at it that the checksum step is being skipped so I'd be happy to take any suggestions regarding this.

Reply:
OK I see what's wrong (as you point out, this behavior predates the changes here, but the new stricter symlink rules uncover this isssue): that test should be updated so that the existing archive is created in one directory, and the stage is created in another i.e. replace

```
Stage(fetcher, path=testpath)
```

with

```
Stage(fetcher, path=tmpdir.someothernewpaththatisnot_testpath)
```